Readme.md
# transducer

[![Build Status](https://secure.travis-ci.org/Gozala/transducer.png)](http://travis-ci.org/Gozala/transducer)

Library for filtering signals based on other signal behavior

## API

### takeWhen(input, behavior)

Takes two reducible signals and returns narrowed down `input` based on
`behavior`. Items form `input` propagate into returned reducible, when
`behavior` yields / becomes `true`-thy. Items do not propagate when `behavior`
yields / becomes `false`-thy until it switches back again. More simply
function returns signal of events from `input` only when `behavior` is `true`.
Result ends when either `input` or `behavior` ends.

```js
var event = require("dom-reduce/event")
var map = require("dom-reduce/map")

var mousedown = event("mousedown", document.documentElement)
var mouseup = event("mouseup", document.documentElement)
var mousemove = event("mousemove", document.documentElement)

function True() { return true }
function False() { return false }

var pressed = merge([map(mousedown, True), map(mouseup, False)])
var drag = takeWhen(mousemove, pressed)
```

### takeUntil(input, behavior)

Takes two reducible signals and returns narrowed down `input` based on
`behavior`. Items form `input` propagate into returned reducible, when
`behavior` yields / becomes `true`-thy until it yields / becomes `false`-thy,
at while point returned reducible ends. More simply function returns signal
of events from `input` until `behavior` is `false`. Result ends when
`behavior` yields / becomes `true` or either one from `input` & `behavior`
ends.

This is similar to `takeWhen` with difference being that result will end
as soon as behavior switches to `false`.


### dropRepeats(input, assert)

Takes reducible `input` and returns narrowed down version with sequential
repeated values dropped. For example, if a given `input` contains items has
following form `< 1 1 2 2 1 >` then result will have a form of `< 1 2 1 >` by
dropping the values that are the same as the previous value. Function takes
second optional argument `assert` that can be used to compare items. Items
to which `assert` returns true will be dropped.

```js
dropRepeats([1, 2, 2, 3, 4, 4, 4, 4, 5])
// => < 1 2 3 4 5 >

dropRepeats([1, "1", 2, "2", 2, 2, 3, 4, "4"])
// => < 1 "1" 2 "2" 2 3 4 "4" >

dropRepeats([1, "1", 2, "2", 2, 3, 4, "4"], function(a, b) {
  return a == b
})
// => < 1 2 3 4 >
```

### throttle(input, [stamp,] delta)

Function takes reducible `input` (supposedly signal or asynchronous stream)
and `delta` value and returns throttled version back. Items are dropped if
time difference between sends is larger then given `delta`. Optionally
`stamp` function maybe passed in order to manually stamp values on which
differences will be asserted.

```js
var input = delay([1, 2, 3, 4, 5, 6, 7], 25)
throttle(input, 40)
// => < 1 3 5 7 >

throttle(clickEvents, function(event) {
  return event.timeStamp
}, 20)
```

## Install

    npm install transducer


History.md
# Changes

## 0.1.0 / 2013-01-25

  - Document APIs in readme.
  - New `throttle` function.
  - New `dropRepeats` function.

## 0.0.1 / 2012-12-05

  - Initial release


test/index.js
"use strict";

exports["test take when"] = require("./take-when")
exports["test take until"] = require("./take-until")
exports["test drop repeats"] = require("./drop-repeats")
exports["test throttle"] = require("./throttle")

require("test").run(exports)


test/throttle.js
"use strict";

var test = require("reducers/test/util/test")
var delay = require("reducers/delay")
var throttle = require("../throttle")

exports["test throttle using stamping function"] = test(function(assert) {
  var now = Date.now()
  var input = [
    { id: 1, time: now },
    { id: 2, time: now + 20 },
    { id: 3, time: now + 25 },
    { id: 4, time: now + 32 },
    { id: 5, time: now + 47 },
    { id: 6, time: now + 60 },
    { id: 7, time: now + 78 }
  ]
  var actual = throttle(input, function(value) { return value.time }, 20)

  assert(actual, [input[0], input[1], input[4], input[6]],
         "throttle items using stamping function")
})

exports["test async input"] = test(function(assert) {
  var input = [{}, {}, {}, {}, {}, {}, {}]
  var actual = throttle(delay(input, 25), 40)

  assert(actual, [input[0], input[2], input[4], input[6]],
         "throttle items by time")
})


if (require.main === module)
  require("test").run(exports)


test/take-when.js
"use strict";

var test = require("reducers/test/util/test")
var event = require("reducers/test/util/event")

var take = require("reducers/take")
var concat = require("reducers/concat")
var delay = require("reducers/delay")


var end = require("reducible/end")

var takeWhen = require("../take-when")

exports["test input ends sampled"]  = test(function(assert) {
  var stop = event()
  var input = event()
  var behavior = event()

  var sampled = takeWhen(input, behavior)
  var actual = concat(sampled, stop)

  assert(actual, [4, 5, 5, 8], "behavior ends sample")

  input.send(1)
  input.send(2)
  behavior.send(false)
  behavior.send(true)
  behavior.send(true)
  behavior.send(false)
  input.send(3)
  behavior.send(true)
  input.send(4)
  input.send(5)
  input.send(5)
  behavior.send(false)
  input.send(6)
  input.send(7)
  behavior.send(true)
  input.send(8)
  input.send(end)

  assert.ok(input.isReduced, "input is reduced")
  assert.ok(!behavior.isReduced, "behavior is not reduced yet")

  behavior.send(false)

  assert.ok(behavior.isReduced, "behavior is reduced on next yield")

  behavior.send(true)
  input.send(9)
  behavior.send(false)
  input.send(7)
  stop.send(end)
})

exports["test behavior ends sampled"]  = test(function(assert) {
  var stop = event()
  var input = event()
  var behavior = event()

  var sampled = takeWhen(input, behavior)
  var actual = concat(sampled, stop)

  assert(actual, [4, 5, 5, 8], "behavior ends sample")

  input.send(1)
  input.send(2)
  behavior.send(false)
  behavior.send(true)
  behavior.send(true)
  behavior.send(false)
  input.send(3)
  behavior.send(true)
  input.send(4)
  input.send(5)
  input.send(5)
  behavior.send(false)
  input.send(6)
  input.send(7)
  behavior.send(true)
  input.send(8)
  behavior.send(end)

  assert.ok(!input.isReduced, "input is not reduced yet")
  assert.ok(behavior.isReduced, "ended behavior is reduced")

  input.send(9)

  assert.ok(input.isReduced, "input is reduced on next yield")

  behavior.send(true)
  behavior.send(false)
  input.send(7)
  stop.send(end)
})

exports["test early behavior switches (last false)"]  = test(function(assert) {
  var stop = event()
  var input = event()
  var behavior = event()

  var sampled = takeWhen(input, behavior)
  var actual = concat(sampled, stop)

  assert(actual, [4, 5, 6], "values from when behavior was true")

  behavior.send(true)
  behavior.send(false)
  behavior.send(true)
  behavior.send(false)
  input.send(1)
  input.send(2)
  input.send(3)
  behavior.send(true)
  input.send(4)
  input.send(5)
  behavior.send(false)
  behavior.send(true)
  input.send(6)
  input.send(end)

  assert.ok(input.isReduced, "ended input is reduced")
  assert.ok(!behavior.isReduced, "behavior is not reduced yet")

  behavior.send(false)

  assert.ok(behavior.isReduced, "behavior is reduced on next yield")

  behavior.send(true)
  input.send(7)
  stop.send(end)
})

exports["test early behavior switches (last true)"]  = test(function(assert) {
  var stop = event()
  var input = event()
  var behavior = event()

  var sampled = takeWhen(input, behavior)
  var actual = concat(sampled, stop)

  assert(actual, [1, 2, 3, 6], "values from input when behavior is true")

  behavior.send(true)
  behavior.send(false)
  behavior.send(true)
  input.send(1)
  input.send(2)
  input.send(3)
  behavior.send(false)
  behavior.send(false)
  input.send(4)
  input.send(5)
  behavior.send(false)
  behavior.send(true)
  input.send(6)
  input.send(end)

  assert.ok(input.isReduced, "ended input is reduced")
  assert.ok(!behavior.isReduced, "behavior is not reduced yet")

  behavior.send(false)

  assert.ok(behavior.isReduced, "behavior is reduced on next yield")

  behavior.send(true)
  input.send(7)
  stop.send(end)
})

exports["test stop reduction before end"] = test(function(assert) {
  var input = event()
  var behavior = event()
  var stop = event()

  var samples = take(takeWhen(input, behavior), 4)
  var actual = concat(samples, stop)

  assert(actual, [3, 4, 5, 8], "take ends transduced signal")

  input.send(1)
  input.send(2)
  behavior.send(true)
  input.send(3)
  behavior.send(true)
  input.send(4)
  input.send(5)
  behavior.send(false)
  input.send(5)
  behavior.send(false)
  input.send(6)
  input.send(7)
  behavior.send(true)
  input.send(8)

  assert.ok(input.isReduced, "input is reduced on yield")
  assert.ok(!behavior.isReduced, "behavior is reduced yet")

  behavior.send(false)

  assert.ok(behavior.isReduced, "behavior is reduced on yield")

  behavior.send(true)
  input.send(7)
  behavior.send(false)
  input.send(8)
  stop.send(end)
})

exports["test error in behavior propagates"] = test(function(assert) {
  var boom = Error("Boom!!")

  var input = event()
  var behavior = event()

  var actual = delay(takeWhen(input, behavior))

  assert(actual, {
    error: boom,
    values: [3, 5]
  }, "error propagate to reducer and stops reducibles")

  input.send(1)
  input.send(2)
  behavior.send(true)
  behavior.send(false)
  input.send(3)
  behavior.send(true)
  input.send(3)
  behavior.send(false)
  input.send(4)
  behavior.send(true)
  input.send(5)
  behavior.send(false)
  behavior.send(boom)

  assert.ok(behavior.isReduced, "behavior is closed or error")
  assert.ok(!input.isReduced, "input is not closed yet")

  input.send(6)

  assert.ok(input.isReduced, "input is reduced on next yield")

  behavior.send(true)
  input.send(7)
  behavior.send(false)
})

exports["test error in input propagates"] = test(function(assert) {
  var boom = Error("Boom!!")

  var input = event()
  var behavior = event()

  var actual = delay(takeWhen(input, behavior))

  assert(actual, {
    error: boom,
    values: [2, 3]
  }, "error propagate to reducer and stops reducibles")

  input.send(1)
  behavior.send(true)
  input.send(2)
  behavior.send(false)
  input.send(3)
  behavior.send(true)
  input.send(3)
  behavior.send(false)
  input.send(4)
  behavior.send(true)
  input.send(boom)

  assert.ok(!behavior.isReduced, "behavior is not yet")
  assert.ok(input.isReduced, "input is closed on error")

  behavior.send(false)

  assert.ok(behavior.isReduced, "behavior is reduced on next yield")

  behavior.send(true)
  input.send(6)
  behavior.send(false)
})


if (require.main === module)
  require("test").run(exports)


take-when.js
"use strict";

var filter = require("reducers/filter")
var concat = require("reducers/concat")
var sample = require("sample")

var nil = new String("void sample")

// Makes sample non `nil` sample if behavior state is truthy.
function makeSample(state, value) { 
  return state ? value : nil
}
// Returns `true` if value isn't `nil` sample
function isntNil(value) { return value !== nil }

function takeWhen(input, behavior) {
  /**
  Takes two reducible signals and returns narrowed down `input` based on
  `behavior`. Items form `input` propagate into returned reducible, when
  `behavior` yields / becomes `true`-thy. Items do not propagate when `behavior`
  yields / becomes `false`-thy until it switches back again. More simply
  function returns signal of events from `input` only when `behavior` is `true`.
  Result ends when either `input` or `behavior` ends.
  **/

  // prepend `false` the behavior to make sure it's first one to yield,
  // that way if behavior switches later, last input value won't propagate.
  var samples = sample(concat(false, behavior), input, makeSample)
  return filter(samples, isntNil)
}

module.exports = takeWhen


drop-repeats.js
"use strict";

var reductions = require("reducers/reductions")
var filter = require("reducers/filter")
var map = require("reducers/map")


var ITEM = 0
var EQUAL = 1

function dropRepeats(input, assert) {
  /**
  Takes reducible `input` and returns narrowed down version with sequential
  repeated values dropped. For example, if a given `input` contains items has
  following form `< 1 1 2 2 1 >` then result will have a form of `< 1 2 1 >` by
  dropping the values that are the same as the previous value. Function takes
  second optional argument `assert` that can be used to compare items. Items
  to which `assert` returns true will be dropped.

  ## Examples

      dropRepeats([1, 2, 2, 3, 4, 4, 4, 4, 5])
      // => < 1 2 3 4 5 >

      dropRepeats([1, "1", 2, "2", 2, 2, 3, 4, "4"])
      // => < 1 "1" 2 "2" 2 3 4 "4" >

      dropRepeats([1, "1", 2, "2", 2, 3, 4, "4"], function(a, b) {
        return a == b
      })
      // => < 1 2 3 4 >
  **/
  var states = reductions(input, function(state, item) {
    var equal = assert ? assert(state[ITEM], item) :
                item === state[ITEM]
    return [item, equal]
  }, [{}])
  var updates = filter(states, function(state) { return !state[EQUAL] })
  return map(updates, function(update) { return update[ITEM] })
}

module.exports = dropRepeats


throttle.js
"use strict";

var reductions = require("reducers/reductions")
var filter = require("reducers/filter")
var map = require("reducers/map")
var dropRepeats = require("./drop-repeats")


var ITEM = 1
var STAMP = 0

function throttle(input, stamp, delta) {
  /**
  Function takes reducible `input` (supposedly signal or asynchronous stream)
  and `delta` value and returns throttled version back. Items are dropped if
  time difference between sends is larger then given `delta`. Optionally
  `stamp` function maybe passed in order to manually stamp values on which
  differences will be asserted.

  ## Examples

      var input = delay([1, 2, 3, 4, 5, 6, 7], 25)
      throttle(input, 40)
      // => < 1 3 5 7 >

      throttle(clickEvents, function(event) {
        return event.timeStamp
      }, 20)
  **/
  if (delta === void(0)) {
    delta = stamp
    stamp = Date.now
  }
  var states = reductions(input, function(state, item) {
     var now = stamp(item)
     return now - state[STAMP] >= delta ? [now, item] :
            state
  }, [0])
  return map(dropRepeats(states), function(state) { return state[ITEM] })
}

module.exports = throttle
